src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -532,6 +544,142 @@ public static int GetCurrentProcessorId()
+        // Check more than once - to make sure it was not because TLS was delayed by GC or a context switch.
+        private static bool SimpleCoreIdSpeedCheck()
+        {
+            // warm up the code paths.
+            int id = UninlinedThreadStatic() | GetCurrentProcessorNumber();
+            long _05usec = Stopwatch.Frequency / 1000000;
+
+            // limit quick test to 100 microseconds.
+            // If we are on slow hardware, we should calibrate anyways.
+            long limit = Stopwatch.Frequency / 10000 + Stopwatch.GetTimestamp();
+            for (int i = 0; i < 10; i++)
+            {
+                int iters = 1;
+                long t1 = 0;
+                // double the sample size until it is 0.5 usec.
+                while (t1 < _05usec)
+                {
+                    iters *= 2;
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        id = GetCurrentProcessorNumber();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                }
+
+                // assuming TLS cannot be a lot slower than getting ID, this should take 1-5 usec
+                long t2 = Stopwatch.GetTimestamp();
+                for (int j = 0; j < iters; j++)
+                {
+                    UninlinedThreadStatic();
+                }
+                long t3 = Stopwatch.GetTimestamp();
+
+                // if getting ID took longer than 2x TLS access, we should consider caching.
+                if (t3 > limit || (t3 - t2) * 2 < t1)
+                {
+                    return false;
+                }
+            }
+
+            // Make sure the result was not negative, which would indicate "Not Supported"
+            return id >= 0;
+        }
+
+        // NoInlining is to make sure JIT does not CSE and to have a better perf proxy for TLS access.
+        // Measuring inlined ThreadStatic in a loop results in underestimates and unnecessary caching.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static int UninlinedThreadStatic()
+        {
+            return CoreIdCache.t_currentProcessorId;
+        }
+
+        // We must collect multiple samples to account for irregularities caused by GC and context switches.
+        // Why we keep an array of samples and do not adjust as we go:
+        //   We expect that we will adjust the refresh rate down. If we do that early, we may end up forcing all
+        //   the calibration work to happen much sooner. There is no urgency in being calibrated while the app
+        //   is in start-up mode. That would just add to the "rush hour" traffic.
+        private static int sCalibrationToDo;
+        private static int sCalibrationDone;
+        // 25 is chosen to budget the sampling under 100 msec total, assuming 4 msec per sample.
```

Comment:
This feels like a lot. I would think a reasonable max budget would be <5ms.

Reply:
It is not contiguous 100msec. It is a bunch of samples each shorter than 5msec. It is hard to make samples shorter though. This is the case where we handle slow APIs, slow machines, coarse or jittery clocks etc.

I do not remember the math right away , but fitting a single sample into < 1msec felt risky of getting unreliable results.

---
In reply to: [352954318](https://github.com/dotnet/runtime/pull/467#discussion_r352954318) [](ancestors = 352954318)

Reply:
number of samples could be less than 25.  I am being conservative, maybe too much,  - in case we get a lot of phony samples due to context switches or GC. 
I did not observe it as a problem, but I have relatively fast hardware and can't rule out this be a bigger problem on something embedded or like that. 

We only need one good measurement of CoreID and TLS between all samples. It feels like 5 samples is too few though. It is possible to be unlucky and get all 5 samples bad on either CoreID or TLS side. 10 samples may be ok..

---
In reply to: [352996637](https://github.com/dotnet/runtime/pull/467#discussion_r352996637) [](ancestors = 352996637,352954318)

Reply:
So what is the worst case regression, e.g. for program that just calls `GetCurrentProcessorId` (e.g. via ArrayPool) right number of times and does not do much of anything else?

Reply:
> I do not remember the math right away , but fitting a single sample into < 1msec felt risky of getting unreliable results.

I would be curious about the reasoning that makes it risky.

Reply:
To time a calibration loop it should take at least a few QPC ticks. Otherwise quantization error may be too high.
Let's say we are ok with 20% error. That puts the long loop at 5 ticks. 

On a typical machine QPC resolution is 10MHz (HyperV is always 10MHz) so tick is 100ns.
So a single loop will need to be at least 0.5msec. 
We need that twice - to find the number of iterations + the actual loop + run the same number of iterations with TLS. 
Let's assume TLS is not worse. It is unlikely to be in calibration if CoreId is faster than TLS.

So the whole calibration cycle takes between 1 and 2 msec. 
(1 if TLS is insignificantly cheap, and 2 if CoreId and TLS are roughly the same)

However that is assuming good hardware.
The effective timer precision is MAX(resolution, access time) Source: https://docs.microsoft.com/en-us/windows/win32/sysinfo/acquiring-high-resolution-time-stamps. 
And we may be dealing with substandard hardware here with either bad resolution or access time or both. 
And it could be on 1GHz CPU. Who knows.

So I doubled the loop time to 1msec. - (10 ticks on decent hardware or HuperV).
And that would put a single calibration sample in 2-4 msec range.

We may consider reducing the number of samples to say 10, but reducing an individual sample even by half seems risky.

Reply:
Am I confusing microseconds and milliseconds here?

Reply:
I think I am. Yes we can fit this under 5 msec total.

Reply:
We are budgeting the entire calibration under 5msec now, if we get to it, so I am resolving this